shared/utility/actuation/InverseKinematics.hpp

```diff
@@ -42,177 +42,143 @@ namespace utility::actuation::kinematics {
-
-        ankleRoll = std::atan2(ankleX.dot(legPlaneTangent), ankleX.dot(hipX));
-
-        const bool isAnkleAboveWaist = unitTargetLeg.dot(Eigen::Matrix<Scalar, 3, 1>::UnitZ()) < 0;
-
-        const Scalar cosZandHipX   = Eigen::Matrix<Scalar, 3, 1>::UnitZ().dot(hipX);
-        const bool hipRollPositive = cosZandHipX <= 0;
-        Eigen::Matrix<Scalar, 3, 1> legPlaneGlobalZ =
-            (isAnkleAboveWaist ? -1 : 1) * (Eigen::Matrix<Scalar, 3, 1>::UnitZ() - (cosZandHipX * hipX));
-        const Scalar legPlaneGlobalZLength = legPlaneGlobalZ.norm();
-        if (legPlaneGlobalZLength > 0) {
-            legPlaneGlobalZ /= legPlaneGlobalZLength;
-        }
-
-        const Scalar cosHipRoll = legPlaneGlobalZ.dot(Eigen::Matrix<Scalar, 3, 1>::UnitZ());
-        hipRoll = (hipRollPositive ? 1 : -1) * std::acos(std::max(std::min(cosHipRoll, Scalar(1)), Scalar(-1)));
-
-        const Scalar phi4 = M_PI - knee - lowerLeg;
-        // Superposition values:
-        const Scalar sinPIminusPhi2 = std::sin(M_PI - phi2);
-        const Eigen::Matrix<Scalar, 3, 1> unitUpperLeg =
-            unitTargetLeg * (std::sin(phi2 - phi4) / sinPIminusPhi2) + ankleY * (std::sin(phi4) / sinPIminusPhi2);
-        const bool isHipPitchPositive = hipX.dot(unitUpperLeg.cross(legPlaneGlobalZ)) >= 0;
-
-        hipPitch = (isHipPitchPositive ? 1 : -1)
-                   * std::acos(std::max(std::min(legPlaneGlobalZ.dot(unitUpperLeg), Scalar(1)), Scalar(-1)));
-
-        // If leg is above waist then hipX is pointing in the wrong direction in the xy plane
-        Eigen::Matrix<Scalar, 3, 1> hipXProjected = (isAnkleAboveWaist ? -1 : 1) * hipX;
-        hipXProjected.z()                         = 0;
-        hipXProjected.normalize();
-        const bool isHipYawPositive = hipXProjected.dot(Eigen::Matrix<Scalar, 3, 1>::UnitY()) >= 0;
-        hipYaw                      = (isHipYawPositive ? 1 : -1)
-                 * std::acos(std::max(std::min(hipXProjected.dot(Eigen::Matrix<Scalar, 3, 1>::UnitX()), Scalar(1)),
-                                      Scalar(-1)));
+        const Scalar cos_hip_roll = leg_plane_global_z.dot(Eigen::Matrix<Scalar, 3, 1>::UnitZ());
+        const Scalar hip_roll =
+            (hip_roll_positive ? 1 : -1) * std::acos(std::max(std::min(cos_hip_roll, Scalar(1)), Scalar(-1)));
+
+        // Hip pitch
+        const Scalar phi4               = knee_pitch - lower_leg;
+        const Scalar sin_pi_minus_phi_2 = std::sin(M_PI - phi2);
+        const Eigen::Matrix<Scalar, 3, 1> unit_upper_leg =
+            unit_target_leg * (std::sin(phi2 - phi4) / sin_pi_minus_phi_2)
+            + ankle_y * (std::sin(phi4) / sin_pi_minus_phi_2);
+        const bool is_hip_pitch_positive = hip_x.dot(unit_upper_leg.cross(leg_plane_global_z)) >= 0;
+        const Scalar hip_pitch =
+            (is_hip_pitch_positive ? -1 : 1)
+            * std::acos(std::max(std::min(leg_plane_global_z.dot(unit_upper_leg), Scalar(1)), Scalar(-1)));
```

Comment:
If you're refactoring, instead of using acos of a dot product it's much more numerically stable to use the "angle between" utility function

Reply:
Wait you don't have that function!
You need this! https://github.com/NUbots/NUbots/pull/1250